examples/pytorch/ogb/directional_GSN/main.py

```diff
@@ -0,0 +1,638 @@
+
+    optimizer = optim.Adam(model.parameters(), lr=params['init_lr'], weight_decay=params['weight_decay'])
+    scheduler = optim.lr_scheduler.ReduceLROnPlateau(optimizer, mode='min',
+                                                     factor=params['lr_reduce_factor'],
+                                                     patience=params['lr_schedule_patience'],
+                                                     verbose=True)
+
+    epoch_train_losses, epoch_val_losses = [], []
+    epoch_train_APs, epoch_val_APs, epoch_test_APs = [], [], []
+
+    train_loader = DataLoader(trainset, batch_size=params['batch_size'], shuffle=True, collate_fn=dataset.collate, pin_memory=True)
+    val_loader = DataLoader(valset, batch_size=params['batch_size'], shuffle=False, collate_fn=dataset.collate, pin_memory=True)
+    test_loader = DataLoader(testset, batch_size=params['batch_size'], shuffle=False, collate_fn=dataset.collate, pin_memory=True)
+
+    with tqdm(range(params['epochs']), unit='epoch') as t:
+        for epoch in t:
+            t.set_description('Epoch %d' % epoch)
+
+            start = time.time()
+
+            epoch_train_loss, epoch_train_ap, optimizer = train_epoch(model, optimizer, device, train_loader)
+            epoch_val_loss, epoch_val_ap = evaluate_network(model, device, val_loader)
+
+            epoch_train_losses.append(epoch_train_loss)
+            epoch_val_losses.append(epoch_val_loss)
+            epoch_train_APs.append(epoch_train_ap.item())
+            epoch_val_APs.append(epoch_val_ap.item())
+
+            _, epoch_test_ap = evaluate_network(model, device, test_loader)
+
+            epoch_test_APs.append(epoch_test_ap.item())
+
+            t.set_postfix(
+                            train_loss=epoch_train_loss, 
+                            train_AP=epoch_train_ap.item(), val_AP=epoch_val_ap.item(),
+                            refresh=False)
+
+            per_epoch_time.append(time.time() - start)
+
+            scheduler.step(-epoch_val_ap.item())
+
+            if optimizer.param_groups[0]['lr'] < params['min_lr']:
+                print("\n!! LR EQUAL TO MIN LR SET.")
+                break
+
+            print('')
+
+    best_val_epoch = np.argmax(np.array(epoch_val_APs))
+    best_train_epoch = np.argmax(np.array(epoch_train_APs))
+    best_val_ap = epoch_val_APs[best_val_epoch]
+    best_val_test_ap = epoch_test_APs[best_val_epoch]
+    best_val_train_ap = epoch_train_APs[best_val_epoch]
+    best_train_ap = epoch_train_APs[best_train_epoch]
+
+    print("Best Train AP: {:.4f}".format(best_train_ap))
+    print("Best Val AP: {:.4f}".format(best_val_ap))
+    print("Test AP of Best Val: {:.4f}".format(best_val_test_ap))
+    print("Train AP of Best Val: {:.4f}".format(best_val_train_ap))
+    print("TOTAL TIME TAKEN: {:.4f}s".format(time.time() - t0))
+    print("AVG TIME PER EPOCH: {:.4f}s".format(np.mean(per_epoch_time)))
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser()
```

Comment:
Users will not need to touch most of these arguments, so perhaps you can simply hard-code most of them when they are used.